### PR TITLE
Add test coverage for `OccurrenceConstraint`

### PR DESCRIPTION
**Pull request by eNeRGy164**
Add test edge case coverage for `OccurrenceConstraint` types

## IMPORTANT 

* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).

### CODE REVIEW

**Review by dennisdoomen — APPROVED** [review]
Muchos gracias.

**Review by dennisdoomen — APPROVED** [review]

**Comment by jnyrup** [comment]
It feels like were getting a little too far away from testing the behavior of the public API.
I think it'll be better if we test this through e.g. `StringAssertions.Contains`.
Could we just modify add assertions to `StringAssertionSpecs.cs` instead?

**Comment by eNeRGy164** [comment]
Yeah, it could also go there.

My thought was that if you go with that, you should also add them in `StringAssertions.ContainEquivalentOf`.
If somebody chooses to add `OccurrenceConstraint` to another assertion method, they would also have to add all 31 possibilities.

So this did look more scalable to me, as `AtMost.Twice()` does not depend on any actual implementation, only the comparison and the value.

**Comment by dennisdoomen** [comment]
Although you know my perspective on testing, I do think the `OccurrenceConstraint` can be seen as a utility function that is designed to be reusable for people building extensions. And then it kind of makes sense to test them through its consumer `AssertionScope`.

**Comment by eNeRGy164** [comment]
@jnyrup If I read Dennis's comment correctly, no changes are required for this PR?

File: Tests/FluentAssertions.Specs/OccurrenceConstraintSpecs.cs [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+            // Arrange
+            OccurrenceConstraint constraint = AtLeast.Once();
+
+            // Act
+            Action act = () => Execute.Assertion
+                .ForConstraint(constraint, actual)
+                .FailWith($"Expected occurrence to be {constraint.Mode} {constraint.ExpectedCount}, but it was {actual}");
+
+            // Assert
+            if (shouldThrow)
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/OccurrenceConstraintSpecs.cs:26]: I don't think we should be compressing tests by introducing conditional logic.
It's only an additional five lines to split the test into a positive and negative tests.

```cs
[InlineData(1)]
[InlineData(2)]
[Theory]
public void At_least_once_is_bigger_than_zero(int occurrences)
{
    // Arrange
    OccurrenceConstraint constraint = AtLeast.Once();

    // Act / Assert
    Execute.Assertion
        .ForConstraint(constraint, occurrences)
        .FailWith("");
}

[Fact]
public void At_least_once_is_not_bigger_than_zero()
{
    // Arrange
    OccurrenceConstraint constraint = AtLeast.Once();

    // Act
    Action act = () => Execute.Assertion
        .ForConstraint(constraint, 0)
        .FailWith($"Expected occurrence to be {constraint.Mode} {constraint.ExpectedCount}, but it was 0");

    // Assert
    act.Should().Throw<XunitException>()
            .WithMessage("Expected occurrence to be at least 1, but it was 0*");
}
```

If we _really_ wanted to compress tests we could cover all scenarios with this (modulo the relaxed failure message assertions)

```cs
public static object[][] PassingConstraints() => new object[][]
{
    new object[] { AtLeast.Once(), 1 },
    new object[] { AtLeast.Once(), 2 },
    new object[] { AtLeast.Twice(), 2 },
    new object[] { AtLeast.Twice(), 3 },
    new object[] { AtLeast.Thrice(), 3 },
    new object[] { AtLeast.Thrice(), 4 },
    new object[] { AtLeast.Times(4), 4 },
    new object[] { AtLeast.Times(4), 5 },

    new object[] { AtMost.Once(), 0 },
    new object[] { AtMost.Once(), 1 },
    new object[] { AtMost.Twice(), 1 },
    new object[] { AtMost.Twice(), 2 },
    new object[] { AtMost.Thrice(), 2 },
    new object[] { AtMost.Thrice(), 3 },
    new object[] { AtMost.Times(4), 3 },
    new object[] { AtMost.Times(4), 4 },

    new object[] { Exactly.Once(), 1 },
    new object[] { Exactly.Twice(), 2 },
    new object[] { Exactly.Thrice(), 3 },
    new object[] { Exactly.Times(4), 4 },

    new object[] { LessThan.Twice(), 1 },
    new object[] { LessThan.Thrice(), 2 },
    new object[] { LessThan.Times(4), 3 },

    new object[] { MoreThan.Once(), 2 },
    new object[] { MoreThan.Twice(), 3 },
    new object[] { MoreThan.Thrice(), 4 },
    new object[] { MoreThan.Times(4), 5 },
};

[Theory]
[MemberData(nameof(PassingConstraints))]
public void Occurrence_constraint_passes(OccurrenceConstraint constraint, int occurrences)
{
    // Act / Assert
    Execute.Assertion
        .ForConstraint(constraint, occurrences)
        .FailWith("");
}

public static object[][] FailingConstraints() => new object[][]
{
    new object[] { AtLeast.Once(), 0 },
    new object[] { AtLeast.Twice(), 1 },
    new object[] { AtLeast.Thrice(), 2 },
    new object[] { AtLeast.Times(4), 3 },

    new object[] { AtMost.Once(), 2 },
    new object[] { AtMost.Twice(), 3 },
    new object[] { AtMost.Thrice(), 4 },
    new object[] { AtMost.Times(4), 5 },

    new object[] { Exactly.Once(), 0 },
    new object[] { Exactly.Once(), 2 },
    new object[] { Exactly.Twice(), 1 },
    new object[] { Exactly.Twice(), 3 },
    new object[] { Exactly.Thrice(), 2 },
    new object[] { Exactly.Thrice(), 4 },
    new object[] { Exactly.Times(4), 3 },
    new object[] { Exactly.Times(4), 5 },

    new object[] { LessThan.Twice(), 2 },
    new object[] { LessThan.Twice(), 3 },
    new object[] { LessThan.Thrice(), 3 },
    new object[] { LessThan.Thrice(), 4 },
    new object[] { LessThan.Times(4), 4 },
    new object[] { LessThan.Times(4), 5 },

    new object[] { MoreThan.Once(), 0 },
    new object[] { MoreThan.Once(), 1 },
    new object[] { MoreThan.Twice(), 1 },
    new object[] { MoreThan.Twice(), 2 },
    new object[] { MoreThan.Thrice(), 2 },
    new object[] { MoreThan.Thrice(), 3 },
    new object[] { MoreThan.Times(4), 3 },
    new object[] { MoreThan.Times(4), 4 },
};

[Theory]
[MemberData(nameof(FailingConstraints))]
public void Occurrence_constraint_fails(OccurrenceConstraint constraint, int occurrences)
{
    // Act
    Action act = () => Execute.Assertion
        .ForConstraint(constraint, occurrences)
        .FailWith($"Expected occurrence to be {constraint.Mode} {constraint.ExpectedCount}, but it was {occurrences}");

    // Assert
    act.Should().Throw<XunitException>()
            .WithMessage("Expected occurrence to be *, but it was *");
}
```

Comment (eNeRGy164) [reply Tests/FluentAssertions.Specs/OccurrenceConstraintSpecs.cs:26]: I really like the `MemberData` example. If @dennisdoomen approves, I will modify it into this; otherwise I will split all the tests.

Comment (dennisdoomen) [reply Tests/FluentAssertions.Specs/OccurrenceConstraintSpecs.cs:26]: I think this is one of those cases where it helps.

**Review by jnyrup — APPROVED** [review]
